### PR TITLE
Fix Race Condition in Parallel AWS Lambda Server Mode Tests

### LINKED ISSUE #5642: Intermittent CI Failures
> FAILED tests/test_awslambda/test_lambda_eventsourcemapping.py::test_invoke_function_from_sqs_exception - botocore.errorfactory.EntityAlreadyExistsException: An error occurred (EntityAlreadyExists) when calling the CreateRole operation: Role with name my-role already exists.

Looks like this started happening after the Server Mode tests for `awslambda` began running in parallel (#4368).

<details><summary>Detailed Log Output</summary>

=================================== FAILURES ===================================
___________________ test_invoke_function_from_sqs_exception ____________________
[gw1] linux -- Python 3.8.14 /opt/hostedtoolcache/Python/3.8.14/x64/bin/python

    def get_role_name():
        with mock_iam():
            iam = boto3.client("iam", region_name=_lambda_region)
            try:
>               return iam.get_role(RoleName="my-role")["Role"]["Arn"]

tests/test_awslambda/utilities.py:170: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <botocore.client.IAM object at 0x7f279c4a5eb0>, args = ()
kwargs = {'RoleName': 'my-role'}

    def _api_call(self, *args, **kwargs):
        # We're accepting *args so that we can give a more helpful
        # error message than TypeError: _api_call takes exactly
        # 1 argument.
        if args:
            raise TypeError(
                f"{py_operation_name}() only accepts keyword arguments."
            )
        # The "self" in this scope is referring to the BaseClient.
>       return self._make_api_call(operation_name, kwargs)

/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/botocore/client.py:495: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <botocore.client.IAM object at 0x7f279c4a5eb0>
operation_name = 'GetRole', api_params = {'RoleName': 'my-role'}

    def _make_api_call(self, operation_name, api_params):
        operation_model = self._service_model.operation_model(operation_name)
        service_name = self._service_model.service_name
        history_recorder.record(
            'API_CALL',
            {
                'service': service_name,
                'operation': operation_name,
                'params': api_params,
            },
        )
        if operation_model.deprecated:
            logger.debug(
                'Warning: %s.%s() is deprecated', service_name, operation_name
            )
        request_context = {
            'client_region': self.meta.region_name,
            'client_config': self.meta.config,
            'has_streaming_input': operation_model.has_streaming_input,
            'auth_type': operation_model.auth_type,
        }
        endpoint_url, additional_headers = self._resolve_endpoint_ruleset(
            operation_model, api_params, request_context
        )
        request_dict = self._convert_to_request_dict(
            api_params=api_params,
            operation_model=operation_model,
            endpoint_url=endpoint_url,
            context=request_context,
            headers=additional_headers,
        )
        resolve_checksum_context(request_dict, operation_model, api_params)
    
        service_id = self._service_model.service_id.hyphenize()
        handler, event_response = self.meta.events.emit_until_response(
            'before-call.{service_id}.{operation_name}'.format(
                service_id=service_id, operation_name=operation_name
            ),
            model=operation_model,
            params=request_dict,
            request_signer=self._request_signer,
            context=request_context,
        )
    
        if event_response is not None:
            http, parsed_response = event_response
        else:
            apply_request_checksum(request_dict)
            http, parsed_response = self._make_request(
                operation_model, request_dict, request_context
            )
    
        self.meta.events.emit(
            'after-call.{service_id}.{operation_name}'.format(
                service_id=service_id, operation_name=operation_name
            ),
            http_response=http,
            parsed=parsed_response,
            model=operation_model,
            context=request_context,
        )
    
        if http.status_code >= 300:
            error_code = parsed_response.get("Error", {}).get("Code")
            error_class = self.exceptions.from_code(error_code)
>           raise error_class(parsed_response, operation_name)
E           botocore.errorfactory.NoSuchEntityException: An error occurred (NoSuchEntity) when calling the GetRole operation: Role my-role not found

/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/botocore/client.py:914: NoSuchEntityException

During handling of the above exception, another exception occurred:

    @pytest.mark.network
    @mock_logs
    @mock_lambda
    @mock_sqs
    def test_invoke_function_from_sqs_exception():
        function_name = str(uuid4())[0:6]
        logs_conn = boto3.client("logs", region_name="us-east-1")
        sqs = boto3.resource("sqs", region_name="us-east-1")
        queue = sqs.create_queue(QueueName=f"{function_name}_queue")
    
        conn = boto3.client("lambda", region_name="us-east-1")
        func = conn.create_function(
            FunctionName=function_name,
            Runtime="python2.7",
>           Role=get_role_name(),
            Handler="lambda_function.lambda_handler",
            Code={"ZipFile": get_test_zip_file_error()},
            Description="test lambda function",
            Timeout=3,
            MemorySize=128,
            Publish=True,
        )

tests/test_awslambda/test_lambda_eventsourcemapping.py:223: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_awslambda/utilities.py:172: in get_role_name
    return iam.create_role(
/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/botocore/client.py:495: in _api_call
    return self._make_api_call(operation_name, kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <botocore.client.IAM object at 0x7f279c4a5eb0>
operation_name = 'CreateRole'
api_params = {'AssumeRolePolicyDocument': 'some policy', 'Path': '/my-path/', 'RoleName': 'my-role'}

    def _make_api_call(self, operation_name, api_params):
        operation_model = self._service_model.operation_model(operation_name)
        service_name = self._service_model.service_name
        history_recorder.record(
            'API_CALL',
            {
                'service': service_name,
                'operation': operation_name,
                'params': api_params,
            },
        )
        if operation_model.deprecated:
            logger.debug(
                'Warning: %s.%s() is deprecated', service_name, operation_name
            )
        request_context = {
            'client_region': self.meta.region_name,
            'client_config': self.meta.config,
            'has_streaming_input': operation_model.has_streaming_input,
            'auth_type': operation_model.auth_type,
        }
        endpoint_url, additional_headers = self._resolve_endpoint_ruleset(
            operation_model, api_params, request_context
        )
        request_dict = self._convert_to_request_dict(
            api_params=api_params,
            operation_model=operation_model,
            endpoint_url=endpoint_url,
            context=request_context,
            headers=additional_headers,
        )
        resolve_checksum_context(request_dict, operation_model, api_params)
    
        service_id = self._service_model.service_id.hyphenize()
        handler, event_response = self.meta.events.emit_until_response(
            'before-call.{service_id}.{operation_name}'.format(
                service_id=service_id, operation_name=operation_name
            ),
            model=operation_model,
            params=request_dict,
            request_signer=self._request_signer,
            context=request_context,
        )
    
        if event_response is not None:
            http, parsed_response = event_response
        else:
            apply_request_checksum(request_dict)
            http, parsed_response = self._make_request(
                operation_model, request_dict, request_context
            )
    
        self.meta.events.emit(
            'after-call.{service_id}.{operation_name}'.format(
                service_id=service_id, operation_name=operation_name
            ),
            http_response=http,
            parsed=parsed_response,
            model=operation_model,
            context=request_context,
        )
    
        if http.status_code >= 300:
            error_code = parsed_response.get("Error", {}).get("Code")
            error_class = self.exceptions.from_code(error_code)
>           raise error_class(parsed_response, operation_name)
E           botocore.errorfactory.EntityAlreadyExistsException: An error occurred (EntityAlreadyExists) when calling the CreateRole operation: Role with name my-role already exists.

</details>

### PR DESCRIPTION
**Pull request by bpandola**
When the Server Mode tests are run in parallel, it's possible for multiple tests to call `get_role_name` at approximately the same time.  When this happens at the beginning of a test run--before `my-role` has been created--the call to `iam.get_role` can fail for more than one caller, all of whom then attempt to execute `iam.create_role`.  The first call to `iam.create_role` will succeed, but all others will result in an `EntityAlreadyExistsException`, which was not handled by the original code and would result in the test failing.

This updated code wraps the entire process in an infinite loop and adds an exception clause for the `iam.create_role` call, which simply passes, allowing the loop to attempt the `iam.get_role` call again.

An explicit test case has been added, which was failing 100% of the time prior to the included `get_role_name` code changes.  Unfortunately, neither the `boto3` client nor our own `IAMBackend` class are thread-safe, so the test itself also includes an infinite retry loop in order to be deterministic and avoid creating another flaky test that fails intermittently.

Note: Normally, infinite loops like this would be a Bad Idea, but this code is only run as part of the test suite (and not `moto` itself) and I simply encountered too many multithreading edge cases to find a suitable alternative. 

Closes #5642

### CODE REVIEW

File: tests/test_awslambda/utilities.py [hunk 1]
```diff
@@ -166,14 +166,18 @@ def create_invalid_lambda(role):
...
-            return iam.get_role(RoleName="my-role")["Role"]["Arn"]
-        except ClientError:
-            return iam.create_role(
-                RoleName="my-role",
-                AssumeRolePolicyDocument="some policy",
-                Path="/my-path/",
-            )["Role"]["Arn"]
+        while True:
+            try:
+                return iam.get_role(RoleName="my-role")["Role"]["Arn"]
```
Comment (bblommers) [inline tests/test_awslambda/utilities.py:171]: So, I don't really remember why I chose to have a single role, but I wonder if it would be easier to just create a new role each time, with a unique name?
AFAIK there is no actual need to share a single role across all functions - creating a new role, one per function/test, may be a lot cleaner and less error-prone.

Comment (bpandola) [reply tests/test_awslambda/utilities.py:171]: Unfortunately, this helper method is called ~100 times throughout the test suite and numerous tests make assertions based on the `my-role` name.  Additionally, the method returns the role's ARN, which would require some parsing to extract the random name (which, again, is currently assumed to be `my-role` in many tests).

I'm with you that this should ultimately be simplified/refactored, but I don't know if I feel like doing the work at the moment.  The changes in this PR at least have the advantage of only being run as part of the tests (and not moto itself).  I can't stand flaky CI (especially when my PRs end up with a red x as a result 😄 ), but I leave it up to you as to whether or not you think that's worth merging it as is or not.

**Review by bblommers — APPROVED** [review]
